Python/Product/Analysis/LanguageServer/EditorFiles.cs

```diff
@@ -0,0 +1,240 @@
+                    // Expected from version hasn't been seen yet, and there are no resets in this
+                    // change, so enqueue it for later.
+                    _server.TraceMessage($"Deferring changes for {uri} until version {fromVersion} is seen");
+                    lock (_pendingChanges) {
+                        _pendingChanges.Add(@params);
+                    }
+                    return;
+                }
+
+                var toVersion = @params.textDocument.version ?? (fromVersion + changes.Length);
+
+                doc.UpdateDocument(part, new DocumentChangeSet(
+                    fromVersion,
+                    toVersion,
+                    changes.Select(c => new DocumentChange {
+                        ReplacedSpan = c.range.GetValueOrDefault(),
+                        WholeBuffer = !c.range.HasValue,
+                        InsertedText = c.text
+                    })
+                ));
+
+                DidChangeTextDocumentParams? next = null;
+                lock (_pendingChanges) {
+                    var notExpired = _pendingChanges
+                        .Where(p => p.textDocument.version.GetValueOrDefault() >= toVersion)
+                        .OrderBy(p => p.textDocument.version.GetValueOrDefault())
+                        .ToArray();
+
+                    _pendingChanges.Clear();
+                    if (notExpired.Any()) {
+                        next = notExpired.First();
+                        _pendingChanges.AddRange(notExpired.Skip(1));
+                    }
+                }
+                if (next.HasValue) {
+                    DidChangeTextDocument(next.Value, null);
+                }
+            } finally {
+                if (enqueueAction != null) {
+                    _server.TraceMessage($"Applied changes to {uri}");
+                    enqueueAction(doc);
+                }
+            }
+        }
+
+        public void UpdateParseDiagnostics(VersionCookie vc, Uri documentUri) {
+            List<PublishDiagnosticsEventArgs> diags = null;
+
+            lock (_lock) {
+                var last = _parseBufferDiagnostics;
+                var lastPublishedAnalysisVersion = _lastReportedAnalysisDiagnostics?.FirstOrDefault()._version;
+
+                foreach (var kv in vc.GetAllParts(documentUri)) {
+                    var part = _server.ProjectFiles.GetPart(kv.Key);
+                    if (!last.TryGetValue(part, out var lastVersion) || lastVersion.Version < kv.Value.Version || _ignoreDiagnosticsVersion) {
+                        last[part] = kv.Value;
+                        // Verify that this version is newer than the last published one so
+                        // parse diagnostics does not overwrite analysis of the same version.
+                        if (!lastPublishedAnalysisVersion.HasValue || lastPublishedAnalysisVersion.Value < kv.Value.Version) {
```

Comment:
`Diagnostic` has a source property on it. How much of this merging logic is unnecessary if we publish the two sets of diagnostics independently?

Reply:
Just this check I think. Analysis includes parsing diagnostic so it is OK for it to overwrite parser but not the other way around. We could have 'Python parser' and 'Python analysis' sources, but it might be confusing to the user, esp with combination with third party linter.

Reply:
The "confusing to user" might be worth testing, if it's going to let us have simpler code here.

We could label them "Python" (parser) and "Python static analysis" (analysis) instead?

Reply:
OK, done